language/move-binary-format/src/compatibility.rs

```diff
@@ -268,3 +268,89 @@ fn type_parameter_phantom_decl_compatibile(
     // old_type_paramter.is_phantom => new_type_parameter.is_phantom
     !old_type_parameter.is_phantom || new_type_parameter.is_phantom
 }
+
+/// A simpler, and stricter compatibility checker relating to the inclusion of the old module in
+/// the new.
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
+pub enum InclusionCheck {
+    Subset,
+    Equal,
+}
+
+impl InclusionCheck {
+    pub fn check(&self, old_module: &Module, new_module: &Module) -> PartialVMResult<()> {
+        let err = Err(PartialVMError::new(
+            StatusCode::BACKWARD_INCOMPATIBLE_MODULE_UPDATE,
+        ));
+
+        // Module checks
+        if old_module.address != new_module.address
+            || old_module.name != new_module.name
+            || old_module.file_format_version != new_module.file_format_version
+        {
+            return err;
+        }
+
+        // If we're checking exactness we make sure there's an inclusion, and that the size of all
+        // of the tables are the exact same.
+        if (self == &Self::Equal)
+            && (old_module.structs.len() != new_module.structs.len()
+                || old_module.exposed_functions.len() != new_module.exposed_functions.len()
+                || old_module.private_functions.len() != new_module.private_functions.len()
+                || old_module.friends.len() != new_module.friends.len()
+                || old_module.constants.len() != new_module.constants.len())
+        {
+            return err;
+        }
+
+        // Struct checks
+        for (name, old_struct) in &old_module.structs {
+            match new_module.structs.get(name) {
+                Some(new_struct) if old_struct == new_struct => (),
+                _ => {
+                    return err;
+                }
+            };
+        }
+
+        // Function checks
+        for (name, old_func) in old_module
+            .exposed_functions
+            .iter()
+            .chain(old_module.private_functions.iter())
+        {
+            match new_module
+                .exposed_functions
+                .get(name)
+                .or_else(|| new_module.private_functions.get(name))
+            {
+                Some(new_func) if old_func == new_func => (),
+                _ => {
+                    return err;
+                }
+            }
+        }
+
+        // Friend checks
```

Comment:
Structuraly speaking, I see why we have this, but are friendships strictly relevant for this check? (i.e. if friendships change but nothing else changed, has the package meaningfully changed, assuming it passes verification?)

Reply:
Structurally I felt better having these checks, but technically they don't need to be checked. I'll remove.